### PR TITLE
Android demo GPU: updating dependencies and migration to Android X

### PR DESCRIPTION
**Pull request by mspnr**
Aligning Android demo GPU with the tutorial:
https://www.tensorflow.org/lite/performance/gpu#android_with_android_studio

- Updating dependencies to the most actual ones
- Migration to AndroidX

The app is tested and working as expected.

### CODE REVIEW

**Comment by abattery** [comment]
@xunkai55 could you review this PR?

**Comment by gbaned** [comment]
@applikationsprogramvara Can you please sign CLA. Thanks!

**Comment by applikationsprogramvara** [comment]
I've signed the CLA under https://cla.developers.google.com/clas

File: tensorflow/lite/java/demo/app/build.gradle [hunk 1]
```diff
@@ -46,15 +45,15 @@ allprojects {
...
-    implementation 'com.android.support:support-annotations:25.3.1'
-    implementation 'com.android.support:support-v13:25.2.0'
+    implementation 'androidx.appcompat:appcompat:1.2.0'
+    implementation 'androidx.constraintlayout:constraintlayout:2.0.4'
+    implementation 'com.google.android.material:material:1.3.0'
+    implementation 'androidx.annotation:annotation:1.2.0'
+    implementation 'androidx.legacy:legacy-support-v13:1.0.0'
 
     // Build off of nightly TensorFlow Lite
-    implementation 'org.tensorflow:tensorflow-lite:0.0.0-nightly-SNAPSHOT'
```
[the remark this answers is not in the document]
Comment (applikationsprogramvara) [reply tensorflow/lite/java/demo/app/build.gradle:56]: @lintian06 the nightly version of tensorflow has been returned and the versions have been updated once again

**Review by lintian06 — APPROVED** [review]

**Comment by lintian06** [comment]
FYI, the PR cannot be pulled for the failures, for:
(1) Conflicting files: tensorflow/lite/java/demo/app/build.gradle, tensorflow/lite/java/demo/build.gradle
(2) The bazel BUILD error for [//tensorflow/lite/java/demo/app/src/main:TfLiteCameraDemo](https://github.com/tensorflow/tensorflow/blob/b8d2ab43ff06d6a67deba79ddaf33d76bb1064a9/tensorflow/lite/java/demo/app/src/main/BUILD#L8)

**Comment by applikationsprogramvara** [comment]
@lintian06 

> (1) Conflicting files: tensorflow/lite/java/demo/app/build.gradle, tensorflow/lite/java/demo/build.gradle

Conflict has been resolved. I took the version from master.

> (2) The bazel BUILD error for [//tensorflow/lite/java/demo/app/src/main:TfLiteCameraDemo](https://github.com/tensorflow/tensorflow/blob/b8d2ab43ff06d6a67deba79ddaf33d76bb1064a9/tensorflow/lite/java/demo/app/src/main/BUILD#L8)

I don't really understand what this bazel BUILD is doing. Probably it should be also somehow tuned. The change I did is working all right when I import demo app in Android studio. The app is also building and running all right on a real device.

Could you give some advice what bazel BUILD is doing and how build issue can be resolved?

**Comment by lintian06** [comment]
[Bazel](https://www.bazel.build/) is a build tool similar to make/maven etc, and supports cross-language compilation and resolves dependencies. TensorFlow uses it to build all projects, and the build rule is BUILD file related to the path.

You can install [bazelisk](https://github.com/bazelbuild/bazelisk) that determines the bazel version smartly.

```
alias bazel='path/to/bazelisk or bazel'
```

```
bazel build //tensorflow/lite/java/demo/app/src/main:TfLiteCameraDemo  # to build this demo target.
```

**Comment by applikationsprogramvara** [comment]
@lintian06 thank you for the explanation. I used Bazel all right for building TFLite libraries for Android and iOS.

But the current usage of building (?) the Android app in Bazel is puzzling me. What is the purpose? To get a complete APK or a Bundle?

Actually if try to build TfLiteCameraDemo using Bazel using the command:
```
bazel build //tensorflow/lite/java/demo/app/src/main:TfLiteCameraDemo
```

I get the following error:
`ERROR: /Users/user/tensorflow/tensorflow/lite/java/demo/app/src/main/BUILD:8:15: no such package '@androidsdk//com.android.support': BUILD file not found in directory 'com.android.support' of external repository @androidsdk. Add a BUILD file to a directory to mark it as a package. and referenced by '//tensorflow/lite/java/demo/app/src/main:TfLiteCameraDemo'`

This can be partially understandable, because we changed `com.android.support` package to `androidx.appcompat` in https://github.com/tensorflow/tensorflow/pull/48577/commits/06ae16ee03613e285d666d7c88956780c492c826.

I have pretty much the same error if I run the same command in the master branch. Which means that I am still missing some build configuration. And yes, I have my ANDROID_SDK and ANDROID_NDK set and run `./confugure` before the build. I don't think that these paths are critical, because Gradle saves `androidx.appcompat` to a separate cache directory.

I am just wondering how @xunkai55 managed to build it when he made his changes recently: https://github.com/tensorflow/tensorflow/commit/0a1f70f80d1308d388ed9c1d40d246a57104ff74. Can we ask him about it?

**Comment by gbaned** [comment]
@lintian06  Can you please take a look on above comments from @applikationsprogramvara. Thanks!

**Comment by xunkai55** [comment]
Hi,

Sorry for the late reply. That commit 0a1f70f was mainly about doc change and gradle change, so the bazel part was untouched in that CL.

To build master, you need to run `./configure` as you mentioned and correctly configure the Android part. I'd recommend 
SDK=23, build_tool_version=28.0.0, NDK=r19c. You might need to install them locally using Android tools.

To resolve the build failure, it's a different story, because it seems that "androidx.compat" is not yet a dependency of TensorFlow.

**Comment by gbaned** [comment]
@applikationsprogramvara Can you please check @xunkai55's comments and keep us posted ? Thanks!

**Comment by applikationsprogramvara** [comment]
@gbaned I am confused by the comment https://github.com/tensorflow/tensorflow/pull/48577#issuecomment-856185077 , because in the current commit mainly `build.gradle` was changed and Bazel part stayed untouched.

Anyway the logic behind this PR is totally different: it is not about  building the app in some testing environment, but to let the developers use the latest version of Android framework available and not update it manually after pulling the repository.

I would suggest to close the PR without merging on the ground of that testing environment does not support the latest Android framework.

**Comment by gbaned** [comment]
@applikationsprogramvara Shall we close the PR then, please confirm. Thank you.

**Pull request closed without merging**